host/class/cdc/usb_host_cdc_acm/cdc_acm_host.c

```diff
@@ -1158,10 +1157,12 @@ esp_err_t cdc_acm_host_data_tx_blocking(cdc_acm_dev_hdl_t cdc_hdl, const uint8_t
     ESP_GOTO_ON_ERROR(usb_host_transfer_submit(cdc_dev->data.out_xfer), unblock, TAG,);
 
     // Wait for OUT transfer completion
-    taken = xSemaphoreTake((SemaphoreHandle_t)cdc_dev->data.out_xfer->context, pdMS_TO_TICKS(timeout_ms));
+    SemaphoreHandle_t transfer_finished_semaphore = (SemaphoreHandle_t)cdc_dev->data.out_xfer->context;
+    taken = xSemaphoreTake(transfer_finished_semaphore, pdMS_TO_TICKS(timeout_ms));
     if (!taken) {
-        // Reset the endpoint
-        cdc_acm_reset_transfer_endpoint(cdc_dev->dev_hdl, cdc_dev->data.out_xfer);
+        cdc_acm_reset_transfer_endpoint(cdc_dev->dev_hdl, cdc_dev->data.out_xfer); // Resetting the endpoint will cause all in-progress transfers to complete
+        xSemaphoreTake(transfer_finished_semaphore, pdMS_TO_TICKS(1000)); // The transfer callback gives the semaphore, take it back
```

Comment:
```suggestion
        xSemaphoreTake(transfer_finished_semaphore, pdMS_TO_TICKS(timeout_ms)); // The transfer callback gives the semaphore, take it back
```
Do we want to wait here for `timeout_ms` amount, or some fixed value?

Reply:
The `timeout_ms` applies to regular transfer timeout. 

This is timeout of channel halt. This should happen 'immediately'. But because the user might run this from a task that has higher priority than the CDC driver task, we wait until we get the 'channel halted' interrupt.

So I think it should be OK like this

Reply:
I think taking the semaphore with 0 timeout just before submit the xfer can avoid this kind of confusion,  also more bug proof.

Reply:
OK, changed!